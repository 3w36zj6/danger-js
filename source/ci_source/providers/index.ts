import { Bitrise } from "./Bitrise"
import { BuddyBuild } from "./BuddyBuild"
import { Buildkite } from "./Buildkite"
import { Circle } from "./Circle"
import { Codeship } from "./Codeship"
import { DockerCloud } from "./DockerCloud"
import { Drone } from "./Drone"
import { FakeCI } from "./Fake"
import { Jenkins } from "./Jenkins"
import { Nevercode } from "./Nevercode"
import { Semaphore } from "./Semaphore"
import { Surf } from "./Surf"
import { TeamCity } from "./TeamCity"
import { Travis } from "./Travis"
import { VSTS } from "./VSTS"
import { Screwdriver } from "./Screwdriver"

const providers = [
  Travis,
  Circle,
  Semaphore,
  Nevercode,
  Jenkins,
  FakeCI,
  Surf,
  DockerCloud,
  Codeship,
  Drone,
  Buildkite,
  BuddyBuild,
  VSTS,
  Bitrise,
  TeamCity,
  Screwdriver,
]

// Mainly used for Dangerfile linting
const realProviders = [
  Travis,
  Circle,
  Semaphore,
  Nevercode,
  Jenkins,
  Surf,
  DockerCloud,
  Codeship,
  Drone,
  Buildkite,
  BuddyBuild,
  VSTS,
  TeamCity,
  Screwdriver,
]

export { providers, realProviders }
