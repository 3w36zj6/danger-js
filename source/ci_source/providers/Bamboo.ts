import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

/**
 * ### CI Setup
 *
 * Bamboo v6.9+ provides the pull request key as an environment variable if
 * the Bamboo branch is created from a pull request.
 * However, Bamboo plan branches created as a result of a new Git branch
 * (or manually) will not have this context added for Danger to utilise:
 *
 * - Go to Plan Configuration -> Branches -> Select:
 *   - Manually
 *   - [x] **When pull request is created**
 *   - When new branch in repository is created
 *   - When new branch in repository is created and matches expression
 *
 * If you would like Bamboo to process all Git branches but still have Danger
 * function on pull requests, a simple workaround is to create a separate Bamboo
 * build plan for pull requests.
 *
 * A more complicated method could be to create a script to consult your source
 * control's API for relevant pull requests based on a branch name, and then
 * invoke Danger for each by setting the `DANGER_MANUAL_CI`,
 * `DANGER_MANUAL_GH_REPO` and `DANGER_MANUAL_PR_NUM` environment variables.
 *
 * ### Token Setup
 *
 * Set the relevant environment variables for your source control system, such
 * as `DANGER_BITBUCKETSERVER_HOST`, `DANGER_BITBUCKETSERVER_USERNAME`, and
 * `DANGER_BITBUCKETSERVER_TOKEN`.
 */

export class Bamboo implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "Bamboo"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["bamboo_buildPlanName", "bamboo_planRepository_repositoryUrl"])
  }

  get isPR(): boolean {
    const mustHave = ["bamboo_repository_pr_key", "bamboo_planRepository_repositoryUrl", "bamboo_buildPlanName"]
    const mustBeInts = ["bamboo_repository_pr_key"]
    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, mustBeInts)
  }

  get pullRequestID(): string {
    return `${this.env.bamboo_repository_pr_key}`
  }

  get repoSlug(): string {
    //ssh://git@bt01.cliplister.com:7999/clfr30/bc3_complete.git
    // bamboo_inject_slug="projects/CLFR30/repos/bc3_complete" \

    const [, project, slug] = this.env.bamboo_planRepository_repositoryUrl.match(
      /\/(\w+)\/([a-zA-Z0-9_\.-]+(?:.git)?)$/
    )

    return `projects/${project}/repos/${slug.replace(/\.[^.]+$/, "")}`
  }
}
