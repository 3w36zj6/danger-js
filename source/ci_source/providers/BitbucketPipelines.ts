import { Env, CISource } from "../ci_source"
import { ensureEnvKeysExist, ensureEnvKeysAreInt } from "../ci_source_helpers"

/**
 * ### CI Setup
 *
 * Bitbucket Pipeline
 *
 * ### Token Setup
 *
 * Add `DANGER_BITBUCKETCLOUD_USERNAME`, `DANGER_BITBUCKETCLOUD_PASSWORD`, and `DANGER_BITBUCKETCLOUD_UUID` to your pipeline repository variable.
 *
 */

export class BitbucketPipelines implements CISource {
  constructor(private readonly env: Env) {}

  get name(): string {
    return "bitbucketPipelines"
  }

  get isCI(): boolean {
    return ensureEnvKeysExist(this.env, ["BITBUCKET_BUILD_NUMBER"])
  }

  get isPR(): boolean {
    const mustHave = ["BITBUCKET_GIT_HTTP_ORIGIN", "BITBUCKET_REPO_OWNER", "BITBUCKET_REPO_SLUG"]
    const mustBeInts = ["BITBUCKET_PR_ID"]
    return ensureEnvKeysExist(this.env, mustHave) && ensureEnvKeysAreInt(this.env, mustBeInts)
  }

  get pullRequestID(): string {
    return this.env.BITBUCKET_PR_ID
  }

  get repoSlug(): string {
    return `${this.env.BITBUCKET_REPO_OWNER}/${this.env.BITBUCKET_REPO_SLUG}`
  }

  get repoURL(): string {
    return this.env.BITBUCKET_GIT_HTTP_ORIGIN
  }
}
