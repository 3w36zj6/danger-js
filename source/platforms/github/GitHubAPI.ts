import { api as fetch } from "../../api/fetch"
import { CISource } from "../../ci_source/ci_source"
import * as find from "lodash.find"

// The Handle the API specific parts of the github

export type APIToken = string

// Note that there are parts of this class which don't seem to be
// used by Danger, they are exposed for Peril support.

/** This represent the GitHub API */

export class GitHubAPI {
  fetch: typeof fetch
  additionalHeaders: any

  constructor(public readonly token: APIToken | undefined, public readonly ciSource: CISource) {
    // This allows Peril to DI in a new Fetch function
    // which can handle unique API edge-cases around integrations
    this.fetch = fetch
    this.additionalHeaders = {}
  }

  /**
   * Grabs the contents of an individual file on GitHub
   *
   * @param {string} path path to the file
   * @param {string} [ref] an optional sha
   * @returns {Promise<string>} text contents
   *
   */
  async fileContents(path: string, ref?: string): Promise<string> {
    // Use head of PR (current state of PR) if no ref passed
    if (!ref) {
      const pr = await this.getPullRequestInfo()
      const prJSON = await pr.json()

      ref = prJSON.head.ref
    }
    const fileMetadata = await this.getFileContents(path, ref)
    const data = await fileMetadata.json()
    const buffer = new Buffer(data.content, "base64")
    return buffer.toString()
  }

  // The above is the API for Platform

  async getDangerCommentID(): Promise<number | null> {
    const userID = await this.getUserID()
    const allCommentsResponse = await this.getPullRequestComments()
    const allComments: any[] = await allCommentsResponse.json()
    const dangerComment = find(allComments, (comment: any) => comment.user.id === userID)
    return dangerComment ? dangerComment.id : null
  }

  async updateCommentWithID(id: number, comment: string): Promise<any> {
    const repo = this.ciSource.repoSlug
    return this.patch(`repos/${repo}/issues/comments/${id}`, {}, {
      body: comment
    })
  }

  async deleteCommentWithID(id: number): Promise<any> {
    const repo = this.ciSource.repoSlug
    return this.api(`repos/${repo}/issues/comments/${id}`, {}, {}, "DELETE")
  }

  async getUserID(): Promise<number> {
    const info = await this.getUserInfo()
    return info.id
  }

  postPRComment(comment: string): Promise<any> {
    const repo = this.ciSource.repoSlug
    const prID = this.ciSource.pullRequestID
    return this.post(`repos/${repo}/issues/${prID}/comments`, {}, {
      body: comment
    })
  }

  getPullRequestInfo(): Promise<any> {
    const repo = this.ciSource.repoSlug
    const prID = this.ciSource.pullRequestID
    return this.get(`repos/${repo}/pulls/${prID}`)
  }

  getPullRequestCommits(): Promise<any> {
    const repo = this.ciSource.repoSlug
    const prID = this.ciSource.pullRequestID
    return this.get(`repos/${repo}/pulls/${prID}/commits`)
  }

  async getUserInfo(): Promise<any> {
    const response: any = await this.get("user")
    return await response.json()
  }

  // TODO: This does not handle pagination
  getPullRequestComments(): Promise<any> {
    const repo = this.ciSource.repoSlug
    const prID = this.ciSource.pullRequestID
    return this.get(`repos/${repo}/issues/${prID}/comments`)
  }

  getPullRequestDiff(): Promise<any> {
    const repo = this.ciSource.repoSlug
    const prID = this.ciSource.pullRequestID
    return this.get(`repos/${repo}/pulls/${prID}`, {
      accept: "application/vnd.github.v3.diff"
    })
  }

  getFileContents(path: string, ref?: string): Promise<any> {
    const repo = this.ciSource.repoSlug
    return this.get(`repos/${repo}/contents/${path}?ref=${ref}`, {})
  }

  async getPullRequests(): Promise<any> {
    const repo = this.ciSource.repoSlug
    const res = await this.get(`repos/${repo}/pulls`)
    if (res.ok) {
      return res.json()
    }
    return []
  }

  private api(path: string, headers: any = {}, body: any = {}, method: string): Promise<any> {
    if (this.token !== undefined) {
      headers["Authorization"] = `token ${this.token}`
    }

    return this.fetch(`https://api.github.com/${path}`, {
      method: method,
      body: body,
      headers: {
        "Content-Type": "application/json",
        ...headers,
        ...this.additionalHeaders
      }
    })
  }

  get(path: string, headers: any = {}, body: any = {}): Promise<any> {
    return this.api(path, headers, body, "GET")
  }

  post(path: string, headers: any = {}, body: any = {}): Promise<any> {
    return this.api(path, headers, JSON.stringify(body), "POST")
  }

  patch(path: string, headers: any = {}, body: any = {}): Promise<any> {
    return this.api(path, headers, JSON.stringify(body), "PATCH")
  }
}
