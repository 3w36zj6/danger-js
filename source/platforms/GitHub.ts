import { GitJSONDSL, GitDSL } from "../dsl/GitDSL"
import { GitHubPRDSL, GitHubDSL, GitHubIssue, GitHubAPIPR, GitHubJSONDSL } from "../dsl/GitHubDSL"
import { GitHubAPI } from "./github/GitHubAPI"
import GitHubUtils from "./github/GitHubUtils"
import gitDSLForGitHub from "./github/GitHubGit"

import * as NodeGitHub from "@octokit/rest"
import { Platform, Comment } from "./platform"

/** Handles conforming to the Platform Interface for GitHub, API work is handle by GitHubAPI */

export class GitHub implements Platform {
  name: string

  constructor(public readonly api: GitHubAPI) {
    this.name = "GitHub"
  }

  /**
   * Get the Code Review description metadata
   *
   * @returns {Promise<any>} JSON representation
   */
  getReviewInfo = (): Promise<GitHubPRDSL> => this.api.getPullRequestInfo()

  /**
   * Get the Code Review diff representation
   *
   * @returns {Promise<GitDSL>} the git DSL
   */
  getPlatformGitRepresentation = (): Promise<GitJSONDSL> => gitDSLForGitHub(this.api)

  /**
   * Gets issue specific metadata for a PR
   */

  getIssue = async (): Promise<GitHubIssue> => {
    const issue = await this.api.getIssue()
    return issue || { labels: [] }
  }

  /**
   * Gets inline comments for current PR
   */
  getInlineComments = async (): Promise<Comment[]> => this.api.getPullRequestInlineComments()

  /**
   * Fails the current build, if status setting succeeds
   * then return true.
   */

  updateStatus = async (passed: boolean, message: string, url?: string): Promise<boolean> => {
    const ghAPI = this.api.getExternalAPI()

    const prJSON = await this.api.getPullRequestInfo()
    const ref = prJSON.head
    try {
      await ghAPI.repos.createStatus({
        repo: ref.repo.name,
        owner: ref.repo.owner.login,
        sha: ref.sha,
        state: passed ? "success" : "failure",
        context: process.env["PERIL_INTEGRATION_ID"] ? "Peril" : "Danger",
        target_url: url || "http://danger.systems/js",
        description: message,
      })
      return true
    } catch (error) {
      return false
    }
  }

  /**
   * Returns the `github` object on the Danger DSL
   *
   * @returns {Promise<GitHubDSL>} JSON response of the DSL
   */
  getPlatformDSLRepresentation = async (): Promise<GitHubJSONDSL> => {
    let pr: GitHubPRDSL
    try {
      pr = await this.getReviewInfo()
    } catch {
      process.exitCode = 1
      throw `
        Could not find pull request information,
        if you are using a private repo then perhaps
        Danger does not have permission to access that repo.
      `
    }

    const issue = await this.getIssue()
    const commits = await this.api.getPullRequestCommits()
    const reviews = await this.api.getReviews()
    const requested_reviewers = await this.api.getReviewerRequests()

    const thisPR = this.APIMetadataForPR(pr)
    return {
      issue,
      pr,
      commits,
      reviews,
      requested_reviewers,
      thisPR,
    }
  }

  supportsCommenting() {
    return true
  }

  supportsInlineComments() {
    return true
  }

  /**
   * Returns the response for the new comment
   *
   * @param {string} comment you want to post
   * @returns {Promise<any>} JSON response of new comment
   */
  createComment = (comment: string) => this.api.postPRComment(comment)

  /**
   * Makes an inline comment if possible. If platform can't make an inline comment with given arguments,
   * it returns a promise rejection. (e.g. platform doesn't support inline comments or line was out of diff).
   *
   * @returns {Promise<any>} JSON response of new comment
   */
  createInlineComment = (git: GitDSL, comment: string, path: string, line: number): Promise<any> => {
    if (!this.supportsInlineComments) {
      return new Promise((_resolve, reject) => reject())
    }

    let commitId = git.commits[git.commits.length - 1].sha

    return this.findPositionForInlineComment(git, line, path).then(position => {
      return this.api.postInlinePRComment(comment, commitId, path, position)
    })
  }

  /**
   * Updates an inline comment if possible. If platform can't update an inline comment,
   * it returns a promise rejection. (e.g. platform doesn't support inline comments or line was out of diff).
   *
   * @returns {Promise<any>} JSON response of new comment
   */
  updateInlineComment = (comment: string, commentId: string): Promise<any> => {
    if (!this.supportsInlineComments) {
      return new Promise((_resolve, reject) => reject())
    }

    return this.api.updateInlinePRComment(comment, commentId)
  }

  /**
   * Finds a position in given diff. This is needed for GitHub API, more on the position finder
   * can be found here: https://developer.github.com/v3/pulls/comments/#create-a-comment
   *
   * @returns {Promise<number>} A number with given position
   */
  findPositionForInlineComment = (git: GitDSL, line: number, path: string): Promise<number> => {
    return git.diffForFile(path).then(diff => {
      return new Promise<number>((resolve, reject) => {
        if (diff === undefined) {
          reject()
        }

        let fileLine = 0
        for (let chunk of diff!.chunks) {
          // Search for a change (that is not a deletion). "ln" is for normal changes, "ln2" for additions,
          // thus need to check for either of them
          let index = chunk.changes.findIndex((c: any) => c.type != "del" && (c.ln == line || c.ln2 == line))
          if (index != -1) {
            fileLine += index + 1
            break
          } else {
            fileLine += chunk.changes.length + 1
          }
        }

        resolve(fileLine)
      })
    })
  }

  /**
   * Deletes the main Danger comment, used when you have
   * fixed all your failures.
   *
   * @returns {Promise<boolean>} did it work?
   */
  deleteMainComment = async (dangerID: string): Promise<boolean> => {
    const commentIDs = await this.api.getDangerCommentIDs(dangerID)
    for (let commentID of commentIDs) {
      await this.api.deleteCommentWithID(commentID)
    }

    return commentIDs.length > 0
  }

  /**
   * Deletes an inline comment, used when you have
   * fixed all your failures.
   *
   * @returns {Promise<boolean>} did it work?
   */
  deleteInlineComment = async (id: number): Promise<boolean> => this.api.deleteInlineCommentWithID(id)

  /**
   * Either updates an existing comment, or makes a new one
   *
   * @param {string} newComment string value of comment
   * @returns {Promise<boolean>} success of posting comment
   */
  async updateOrCreateComment(dangerID: string, newComment: string): Promise<boolean> {
    const commentIDs = await this.api.getDangerCommentIDs(dangerID)

    if (commentIDs.length) {
      // Edit the first comment
      await this.api.updateCommentWithID(commentIDs[0], newComment)

      // Delete any dupes
      for (let commentID of commentIDs) {
        if (commentID !== commentIDs[0]) {
          await this.api.deleteCommentWithID(commentID)
        }
      }
    } else {
      await this.createComment(newComment)
    }

    return true
  }

  /**
   * Converts the PR JSON into something easily used by the Github API client.
   */
  APIMetadataForPR = (pr: GitHubPRDSL): GitHubAPIPR => {
    return {
      number: pr.number,
      repo: pr.base.repo.name,
      owner: pr.base.repo.owner.login,
    }
  }

  getFileContents = this.api.fileContents
}

// This class should get un-classed, but for now we can expand by functions

export const githubJSONToGitHubDSL = (gh: GitHubJSONDSL, api: NodeGitHub): GitHubDSL => {
  return {
    ...gh,
    api,
    utils: GitHubUtils(gh.pr, api),
  }
}
