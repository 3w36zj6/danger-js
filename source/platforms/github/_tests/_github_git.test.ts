import { GitHub } from "../../GitHub"
import { GitHubAPI } from "../GitHubAPI"

import { GitCommit } from "../../../dsl/Commit"
import { FakeCI } from "../../../ci_source/providers/Fake"
import { readFileSync } from "fs"
import { resolve } from "path"
import * as os from "os"

const fixtures = resolve(__dirname, "..", "..", "_tests", "fixtures")
const EOL = os.EOL

/** Returns JSON from the fixtured dir */
export const requestWithFixturedJSON = async (path: string): Promise<() => Promise<any>> => () => (
  Promise.resolve(JSON.parse(readFileSync(`${fixtures}/${path}`, {}).toString()))
)

/** Returns arbitrary text value from a request */
export const requestWithFixturedContent = async (path: string): Promise<() => Promise<string>> => () => (
  Promise.resolve(readFileSync(`${fixtures}/${path}`, {}).toString())
)

describe("the dangerfile gitDSL", async () => {
  let github: GitHub = {} as any
  beforeEach(async () => {
    const api = new GitHubAPI(new FakeCI({}))
    github = new GitHub(api)

    api.getPullRequestInfo = await requestWithFixturedJSON("github_pr.json")
    api.getPullRequestDiff = await requestWithFixturedContent("github_diff.diff")
    api.getPullRequestCommits = await requestWithFixturedJSON("github_commits.json")
    api.getFileContents = async (path, repoSlug, ref) => (await requestWithFixturedJSON(`static_file:${ref}.json`))()
  })

  it("sets the modified/created/deleted", async () => {
    const gitDSL = await github.getPlatformGitRepresentation()

    expect(gitDSL.modified_files).toEqual(["CHANGELOG.md", "data/schema.graphql", "data/schema.json", "externals/metaphysics", "lib/__mocks__/react-relay.js", "lib/components/artist/about.js", "lib/components/gene/header.js", "lib/containers/__tests__/__snapshots__/gene-tests.js.snap", "lib/containers/__tests__/gene-tests.js", "lib/containers/gene.js", "tsconfig.json"]) //tslint:disable-line:max-line-length

    expect(gitDSL.created_files).toEqual(["lib/components/gene/about.js", "lib/components/gene/biography.js", "lib/components/related_artists/index.js", "lib/components/related_artists/related_artist.js"]) //tslint:disable-line:max-line-length

    expect(gitDSL.deleted_files).toEqual(["lib/components/artist/related_artists/index.js", "lib/components/artist/related_artists/related_artist.js", "lib/components/gene/about_gene.js"]) //tslint:disable-line:max-line-length
  })

  it("shows the diff for a specific file", async () => {
    const expected = ` - [dev] Updates Flow to 0.32 - orta${EOL} - [dev] Updates React to 0.34 - orta${EOL} - [dev] Turns on "keychain sharing" to fix a keychain bug in sim - orta${EOL}+- GeneVC now shows about information, and trending artists - orta${EOL} ${EOL} ### 1.1.0-beta.2${EOL} ` //tslint:disable-line:max-line-length
    const gitDSL = await github.getPlatformGitRepresentation()
    const {diff} = await gitDSL.diffForFile("CHANGELOG.md")

    expect(diff).toEqual(expected)
  })

  it("should show only diff of specified type", async () => {
    const expected = "+- GeneVC now shows about information, and trending artists - orta"
    const gitDSL = await github.getPlatformGitRepresentation()
    const {diff} = await gitDSL.diffForFile("CHANGELOG.md", ["add"])

    expect(diff).toEqual(expected)
  })

  it("should include `before` text content of the file", async () => {
    const gitDSL = await github.getPlatformGitRepresentation()
    const {before} = await gitDSL.diffForFile("CHANGELOG.md")

    expect(before).toMatchSnapshot()
  })

  it("should include `after` text content of the file", async () => {
    const gitDSL = await github.getPlatformGitRepresentation()
    const {after} = await gitDSL.diffForFile("CHANGELOG.md")

    expect(after).toMatchSnapshot()
  })

  it("sets up commit data correctly", async () => {
    const exampleCommit: GitCommit = {
      "author": {
        "date": "2016-09-30T13:52:14Z",
        "email": "orta.therox@gmail.com",
        "name": "Orta Therox",
      },
      "committer": {
        "date": "2016-09-30T13:52:14Z",
        "email": "orta.therox@gmail.com",
        "name": "Orta Therox",
      },
      "message": "WIP on Gene",
      "parents": ["98f3e73f5e419f3af9ab928c86312f28a3c87475"],
      "sha": "13da2c844def1f4262ee440bd86fb2a3b021718b",
      "tree": {
        "sha": "d1b7448d7409093054efbb06ae12d1ffb002b956",
        "url": "https://api.github.com/repos/artsy/emission/git/trees/d1b7448d7409093054efbb06ae12d1ffb002b956",
      },
      "url": "https://api.github.com/repos/artsy/emission/commits/13da2c844def1f4262ee440bd86fb2a3b021718b"
    }
    const gitDSL = await github.getPlatformGitRepresentation()
    expect(gitDSL.commits[0]).toEqual(exampleCommit)
  })

  describe("JSONPatchForFile", () => {
    it("returns a null for files not in the modified_files", async () => {
      const gitDSL = await github.getPlatformGitRepresentation()
      const empty = await gitDSL.JSONPatchForFile("fuhqmahgads.json")
      expect(empty).toEqual(null)
    })

    it("handles showing a patch for two different diff files", async () => {
      const before = {
        a: "Hello, world",
        b: 1,
        c: ["one", "two", "three"]
      }

      const after = {
        a: "o, world",
        b: 3,
        c: ["one", "two", "three", "four"]
      }

      github.api.fileContents = async (path, repo, ref) => {
        const obj = (ref === "master") ? before : after
        return JSON.stringify(obj)
      }

      const gitDSL = await github.getPlatformGitRepresentation()
      const empty = await gitDSL.JSONPatchForFile("data/schema.json")
      expect(empty).toEqual({
        before,
        after,
        diff: [
          { "op": "replace", "path": "/a", "value": "o, world" },
          { "op": "replace", "path": "/b", "value": 3 },
          { "op": "add", "path": "/c/-", "value": "four" }
        ]
      }
      )
    })
  })

  describe("JSONDiffForFile", () => {
    it("returns an empty object for files not in the modified_files", async () => {
      const gitDSL = await github.getPlatformGitRepresentation()
      const empty = await gitDSL.JSONDiffForFile("fuhqmahgads.json")
      expect(empty).toEqual({})
    })

    it("handles showing a patch for two different diff files", async () => {
      github.api.fileContents = async (path, repo, ref) => {
        const before = {
          a: "Hello, world",
          b: 1,
          c: ["one", "two", "three"], // add
          d: ["one", "two", "three"], // remove
          e: ["one", "two", "three"], // replace
        }

        const after = {
          a: "o, world",
          b: 3,
          c: ["one", "two", "three", "four"],
          d: ["one", "two"],
          e: ["five", "one", "three"]
        }

        const obj = (ref === "master") ? before : after
        return JSON.stringify(obj)
      }
      const gitDSL = await github.getPlatformGitRepresentation()
      const empty = await gitDSL.JSONDiffForFile("data/schema.json")
      expect(empty).toEqual({
        "a": { "after": "o, world", "before": "Hello, world" },
        "b": { "after": 3, "before": 1 },
        "c": { "added": ["four"], "after": ["one", "two", "three", "four"], "before": ["one", "two", "three"], "removed": [] },
        "d": { "added": [], "after": ["one", "two"], "before": ["one", "two", "three"], "removed": ["three"] },
        "e": { "added": ["five"], "after": ["five", "one", "three"], "before": ["one", "two", "three"], "removed": ["two"] }
      })
    })

    it("handles a package.json change elegantly", async () => {
      github.api.fileContents = async (path, repo, ref) => {
        const before = {
          "dependencies": {
            "babel-polyfill": "^6.20.0",
            "chalk": "^1.1.1",
            "commander": "^2.9.0",
            "debug": "^2.6.0"
          },
          "devDependencies": {
            "babel-cli": "^6.16.0",
            "babel-plugin-syntax-async-functions": "^6.13.0",
            "babel-plugin-transform-flow-strip-types": "^6.8.0",
          }
        }

        const after = {
          "dependencies": {
            // "babel-polyfill": "^6.20.0",
            "chalk": "^1.2.1", // from ^1.1.1 to ^1.2.1
            "commander": "^2.9.0",
            "debug": "^2.6.0"
          },
          "devDependencies": {
            "babel-cli": "^6.16.0",
            "babel-plugin-typescript": "^2.2.0", // hah - I wish (added)
            "babel-plugin-syntax-async-functions": "^6.13.0",
            "babel-plugin-transform-flow-strip-types": "^6.8.0",
          }
        }

        const obj = (ref === "master") ? before : after
        return JSON.stringify(obj)
      }
      const gitDSL = await github.getPlatformGitRepresentation()
      const empty = await gitDSL.JSONDiffForFile("data/schema.json")
      expect(empty).toEqual({
        "dependencies": {
          "added": [],
          "after": { "chalk": "^1.2.1", "commander": "^2.9.0", "debug": "^2.6.0" },
          "before": { "babel-polyfill": "^6.20.0", "chalk": "^1.1.1", "commander": "^2.9.0", "debug": "^2.6.0" },
          "removed": ["babel-polyfill"]
        },
         "devDependencies": {
           "added": ["babel-plugin-typescript"],
           "after": { "babel-cli": "^6.16.0", "babel-plugin-syntax-async-functions": "^6.13.0", "babel-plugin-transform-flow-strip-types": "^6.8.0", "babel-plugin-typescript": "^2.2.0" }, //tslint:disable-line:max-line-length
           "before": { "babel-cli": "^6.16.0", "babel-plugin-syntax-async-functions": "^6.13.0", "babel-plugin-transform-flow-strip-types": "^6.8.0" },
           "removed": []
          }
      })
    })
  })
})
