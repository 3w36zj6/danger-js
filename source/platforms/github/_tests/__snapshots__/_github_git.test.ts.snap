// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the dangerfile gitDSL should include \`added\` text content of the file 1`] = `
"+import About from '../components/gene/about'
+    this.state = { selectedTabIndex: 1 }
+      case TABS.ABOUT: return <About gene={this.props.gene} />
+        \${About.getFragment('gene')}"
`;

exports[`the dangerfile gitDSL should include \`after\` text content of the file 1`] = `
"/* @flow */
'use strict'

import Relay from 'react-relay'
import React from 'react'
import { Text, ScrollView, View, Dimensions } from 'react-native'

import Events from '../native_modules/events'

import About from '../components/gene/about'
import Header from '../components/gene/header'

import TabView from '../components/tab_view'
import type TabSelectionEvent from '../components/events'

const TABS = {
  WORKS: 'WORKS',
  ABOUT: 'ABOUT',
}

class Gene extends React.Component {

  state: {
    selectedTabIndex: number,
  };

  componentWillMount() {
    this.state = { selectedTabIndex: 1 }
  }

  tabSelectionDidChange = (event: TabSelectionEvent) => {
    this.setState({ selectedTabIndex: event.nativeEvent.selectedIndex })
  }

  availableTabs = () => {
    return [TABS.WORKS, TABS.ABOUT]
  }

  selectedTabTitle = () => {
    return this.availableTabs()[this.state.selectedTabIndex]
  }

  // This is *not* called on the initial render, thus it will only post events for when the user actually taps a tab.
  componentDidUpdate(previousProps, previousState) {
    Events.postEvent(this, {
      name: 'Tapped gene view tab',
      tab: this.selectedTabTitle().toLowerCase(),
      gene_id: this.props.gene._id,
      gene_slug: this.props.gene.id,
    })
  }

  renderSelectedTab = () => {
    switch (this.selectedTabTitle()) {
      // case TABS.ABOUT: return <About gene={this.props.gene} />
      // case TABS.WORKS: return <Artworks gene={this.props.gene} />
      case TABS.ABOUT: return <About gene={this.props.gene} />
      case TABS.WORKS: return <Text/>
    }
  }

  renderTabView() {
    return (
      <TabView titles={this.availableTabs()} selectedIndex={this.state.selectedTabIndex} onSelectionChange={this.tabSelectionDidChange}>
        { this.renderSelectedTab() }
      </TabView>
    )
  }

  render() {
    const windowDimensions = Dimensions.get('window')
    const commonPadding = windowDimensions.width > 700 ? 40 : 20

    return (
      <ScrollView scrollsToTop={true}>
        <View style={{ paddingLeft: commonPadding, paddingRight: commonPadding }}>
          <Header gene={this.props.gene} />
          { this.renderTabView() }
        </View>
      </ScrollView>
    )
  }
}

export default Relay.createContainer(Gene, {
  fragments: {
    gene: () => Relay.QL\`
      fragment on Gene {
        _id
        id
        \${Header.getFragment('gene')}
        \${About.getFragment('gene')}
      }
    \`,
  }
})
"
`;

exports[`the dangerfile gitDSL should include \`before\` text content of the file 1`] = `
"/* @flow */
'use strict'

import Relay from 'react-relay'
import React from 'react'
import { Text, ScrollView, View, Dimensions } from 'react-native'

import Events from '../native_modules/events'

import Header from '../components/gene/header'

import TabView from '../components/tab_view'
import type TabSelectionEvent from '../components/events'


const TABS = {
  WORKS: 'WORKS',
  ABOUT: 'ABOUT',
}

class Gene extends React.Component {

  state: {
    selectedTabIndex: number,
  };

  componentWillMount() {
    this.state = { selectedTabIndex: 0 }
  }

  tabSelectionDidChange = (event: TabSelectionEvent) => {
    this.setState({ selectedTabIndex: event.nativeEvent.selectedIndex })
  }

  availableTabs = () => {
    return [TABS.WORKS, TABS.ABOUT]
  }

  selectedTabTitle = () => {
    return this.availableTabs()[this.state.selectedTabIndex]
  }

  // This is *not* called on the initial render, thus it will only post events for when the user actually taps a tab.
  componentDidUpdate(previousProps, previousState) {
    Events.postEvent(this, {
      name: 'Tapped gene view tab',
      tab: this.selectedTabTitle().toLowerCase(),
      gene_id: this.props.gene._id,
      gene_slug: this.props.gene.id,
    })
  }

  renderSelectedTab = () => {
    switch (this.selectedTabTitle()) {
      // case TABS.ABOUT: return <About gene={this.props.gene} />
      // case TABS.WORKS: return <Artworks gene={this.props.gene} />
      case TABS.ABOUT: return <Text/>
      case TABS.WORKS: return <Text/>
    }
  }

  renderTabView() {
    return (
      <TabView titles={this.availableTabs()} selectedIndex={this.state.selectedTabIndex} onSelectionChange={this.tabSelectionDidChange}>
        { this.renderSelectedTab() }
      </TabView>
    )
  }

  render() {
    const windowDimensions = Dimensions.get('window')
    const commonPadding = windowDimensions.width > 700 ? 40 : 20

    return (
      <ScrollView scrollsToTop={true}>
        <View style={{ paddingLeft: commonPadding, paddingRight: commonPadding }}>
          <Header gene={this.props.gene} />
          { this.renderTabView() }
        </View>
      </ScrollView>
    )
  }
}

export default Relay.createContainer(Gene, {
  fragments: {
    gene: () => Relay.QL\`
      fragment on Gene {
        _id
        id
        \${Header.getFragment('gene')}
      }
    \`,
  }
})
"
`;

exports[`the dangerfile gitDSL should include \`removed\` text content of the file 1`] = `
"-
-    this.state = { selectedTabIndex: 0 }
-      case TABS.ABOUT: return <Text/>"
`;

exports[`the dangerfile gitDSL shows the diff for a specific file 1`] = `
" 
 import Events from '../native_modules/events'
 
+import About from '../components/gene/about'
 import Header from '../components/gene/header'
 
 import TabView from '../components/tab_view'
 import type TabSelectionEvent from '../components/events'
 
-
 const TABS = {
   WORKS: 'WORKS',
   ABOUT: 'ABOUT',
   };
 
   componentWillMount() {
-    this.state = { selectedTabIndex: 0 }
+    this.state = { selectedTabIndex: 1 }
   }
 
   tabSelectionDidChange = (event: TabSelectionEvent) => {
     switch (this.selectedTabTitle()) {
       // case TABS.ABOUT: return <About gene={this.props.gene} />
       // case TABS.WORKS: return <Artworks gene={this.props.gene} />
-      case TABS.ABOUT: return <Text/>
+      case TABS.ABOUT: return <About gene={this.props.gene} />
       case TABS.WORKS: return <Text/>
     }
   }
         _id
         id
         \${Header.getFragment('gene')}
+        \${About.getFragment('gene')}
       }
     \`,
   }"
`;
